docker/generic/Dockerfile.base

```diff
@@ -0,0 +1,97 @@
+#
+# Install ROS packages used by Autoware.
+#
+ARG ROS_DISTRO
+FROM ros:$ROS_DISTRO
+
+#
+# Install tools and libraries required by Autoware
+#
+RUN apt-get update && apt-get install -y \
+        cmake-curses-gui \
+        cmake-qt-gui \
+        dbus-x11 \
+        dmz-cursor-theme \
+        fonts-dejavu \
+        gconf2 \
+        gnome-terminal \
+        gstreamer0.10-plugins-good \
+        language-pack-en \
+        libarmadillo-dev \
+        libcanberra-gtk-module \
+        libcanberra-gtk3-0 \
+        libcanberra-gtk3-module \
+        libdbus-glib-1-2 \
+        libgflags-dev \
+        libglew-dev \
+        libgoogle-glog-dev \
+        libgoogle-perftools-dev \
+        libgsl0-dev \
+        libmosquitto-dev \
+        libopencv-dev \
+        libopenni2-dev \
+        libpcap-dev \
+        libssh2-1-dev \
+        locales \
+        pulseaudio \
+        python-flask \
+        python-requests \
+        python3-colcon-common-extensions \
+        python3-pip \
+        python3-setuptools \
+        sudo \
+        tmux \
+        v4l-utils \
+        vim \
+        wget && \
+        pip3 install -U setuptools && \
+        rm -rf /var/lib/apt/lists/*
+
+#
+# Configure environmet
+#
+
+RUN update-locale LANG=en_US.UTF-8 LC_MESSAGES=POSIX
+
+# Add user
+ENV USERNAME autoware
+ARG USER_ID=1000
+ARG GROUP_ID=15214
+ENV PULSE_SERVER /run/pulse/native
+
+RUN groupadd --gid $GROUP_ID $USERNAME && \
+        useradd --gid $GROUP_ID -m $USERNAME && \
+        echo "$USERNAME:$USERNAME" | chpasswd && \
+        usermod --shell /bin/bash $USERNAME && \
+        usermod -aG sudo $USERNAME && \
+        usermod  --uid $USER_ID $USERNAME && \
+        echo "$USERNAME ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers.d/$USERNAME && \
+        chmod 0440 /etc/sudoers.d/$USERNAME
+
+# Startup scripts
+ENV LANG="en_US.UTF-8"
+RUN echo "source /opt/ros/$ROS_DISTRO/setup.bash" >> /etc/profile.d/ros.sh && \
+        # Fix for QT and X server errors
+        echo "export QT_X11_NO_MITSHM=1" >> /etc/profile.d/autoware.sh && \
+        # Set defaut language
+        echo "export LANG=\"en_US.UTF-8\"" >> /etc/profile.d/autoware.sh
+
+#
+# Scan Autoware packages to install dependencies
+#
+COPY ./ /tmp/Autoware
+RUN apt-get update && \
+        apt-get install -y ros-$ROS_DISTRO-desktop-full && \
+        rosdep install -y --from-paths /tmp/Autoware/ros/src --ignore-src && \
```

Comment:
I'd advise against using rosdep here as it prevents Docker from using the cache effectively. There's also no trace of what packages are being installed, and more importantly, given that ROS1 does not guarantee ABI compatibility, any  ROS package installed afterwards may cause a segfault.

Reply:
> @amc-nu See below reply to the different points:
> 
> > @filiperinaldi @sgermanserrano Thanks for all the effort put on this PR. Looks great.
> > I have some questions/observations:
> > ```
> > 1. Is there any recommended nvidia-docker version?
> > ```
> 
> There is a need to use nvidia-docker v2
> 

I'll add the version to the README.md.

Reply:
The previous Dockerfile had a list of ROS packages being installed. They were not comprehensive and were very likely to be out of sync as the dependencies change from version to version.

The rationale for doing it like this was that for a given version of Autoware, its corresponding Docker base image will have the dependencies pre-installed. If you are using this based to develop nem features, then you will have to manually run rosdep again to complement the new dependencies required.

The effect on Docker's cache could be minimised with a list, but the cache would still be discarded eventually when the list gets updated. It is a trade-off between getting the dependencies automatically sorted out (instead of someone manually finding them out and updating the Dockerfile) vs increasing the usage of cache.
What do you think would be a better balance? Perhaps splitting ros-desktop-full and a few other basic ROS packages out into their own RUN command to at least re-use that layer? Move the rosdep step to be part of the build steps?

I'm not sure I understand the compatibility issue. Aren't these all deb packages which are meant to be compatible on the same distro?

Reply:
> What do you think would be a better balance? Perhaps splitting ros-desktop-full and a few other basic ROS packages out into their own RUN command to at least re-use that layer? Move the rosdep step to be part of the build steps?

I think having a list of known packages and then running rosdep install is a good compromise. OSRF has the process entirely automated in their images, perhaps in the future we could just have a script that runs rosdep install and if there are any new packages, add them to Dockerfile and commit the changes.

> I'm not sure I understand the compatibility issue. Aren't these all deb packages which are meant to be compatible on the same distro?

Unfortunately not. If I have packages installed from a previous sync and I install new ones, they are most certainly to crash because they are not ABI-compatible. ROS releases must be upgraded/installed all at once, you can't mix packages from different syncs, even if they are from the same distribution.

Reply:
Just a correction, it's not certain as I stated, but there's a chance that the packages are incompatible.

Reply:
Done.